### ISSUE TITLE
package hash for dependency changes if package_id() is changed for header_only library

### ISSUE THREAD

**Issue by sunj1** [conan-io/conan#3471]
- I have a header_only library package, say `lib_head` with package-id, say, `hash1`
- I have a consumer library, say `lib1`, which consumes this package
- I have a binary which consumes the library `lib1` with package-id `lib1-hash1`
- Now, I change the package_id() method of `lib_head` and uploaded it (the new package-id is `hash2`)
- When I try to build the binary, it requires `lib1` with a different package-id `lib1-hash2`.

Why did this package requirement change?

I have verified that all settings and options are same for `lib1` but with a new hash.

- [x] I've read the [CONTRIBUTING guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

**Comment by memsharded** [comment]
Could you please share a reproducible example? What is your ``package_id()`` of lib1?

**Comment by sunj1** [comment]
Sorry if the example was not clear.
I'll to elaborate more.
- I have a header_only library with package_id `hash1`
- I have a library `lib1`, which uses the header_only library package
- I create a binary which uses `lib1`, and requires its package_id to be  `4080ed130005757cda2b32dbf5124ee5f85bf194`
- Now, I define package_id() function in the header_only library and its package_id changes to `hash2`
- Note that I do not re-create the `lib1` library (I do not have access to it)
- When I try to build the binary, it requires `lib1` with a new hash `63e09072d3cc372237bcc3f69136ec680d2585b1`

**Comment by memsharded** [comment]
> Note that I do not re-create the lib1 library (I do not have access to it, and its created with older conan version)

I need to know if ``lib1`` has some ``package_id()`` method defined, and if yes, what is its implementation. Even if you don't recreate the binary, the way it defines its package-ID is implemented in that method. You can inspect the conanfile with ``conan get lib1/version@user/channel``

**Comment by sunj1** [comment]
I checked that `packaged_id()` is never used in any conan files.

**Comment by sunj1** [comment]
The package() method is a little different though, in the **header_only library**

Earlier it was
```
self.copy("*", dst="include", src="include")
```

Now, I changed it in my base conan-class to
```
self.copy("*.h", dst="include", src="include")
self.copy("*.hpp", dst="include", src="include")
self.copy("*.a", dst="lib", src="_build/lib")
```

**Comment by memsharded** [comment]
I am afraid that the issue could be the previous conan version ``0.14.1``. There were some bug pre conan 1.0 regarding the computation of the package ID that were fixed before release 1.0. When you said older version, I thought you were talking about some 1.X one, sorry.

If this is the case, the package-ID of ``lib1`` should probably change even if you don't do changes to header_only lib, just by firing again the build. Could you please verify it? In any case I think the best approach would be to re-generate packages for ``lib1``, as it was generated with a conan version almost 2 years old, and way before conan went out of beta in conan 1.0.

**Comment by sunj1** [comment]
Closing issue, since the computation of package_id has changed as compared to my much-older conan version.
